### ISSUE TITLE
Type inference causing problems when is_dataclass() is tested in decorator

### ISSUE THREAD

**Issue by hmeine** [microsoft/pylance-release#5893]
<h3> Environment data </h3>

-   Language Server version: 2024.5.1
-   OS and version: darwin arm64
-   Python version (and distribution if applicable, e.g. Anaconda): MacPorts, Python 3.11.9
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: basic

<h3> Code Snippet </h3>

The following MWE shows the problem I have with a class decorator I wrote that *never changes the type of the decorated class* (but in reality is just used to register it in some way) which uses `dataclasses.is_dataclass()`.  While minimizing the example, the error changed twice, but I think the underlying problem should still be the same.

```python
import dataclasses
from typing import Type

def decorator1(DataClass: Type, condition: bool = False):
    '''This one breaks the infererred type of Foo (which should be Foo in any case)'''
    if condition and dataclasses.is_dataclass(DataClass):
        pass
    return DataClass

def decorator2(DataClass: Type):
    '''This one seems to change the infererred type from Foo to object'''
    if dataclasses.is_dataclass(DataClass):
        pass
    return DataClass

def decorator3(DataClass: Type):
    '''This one works'''
    return DataClass


@dataclasses.dataclass
class Properties:
    some_prop: int = 42

@decorator1
class Foo:
    def properties(self) -> Properties:
        return Properties()

class Bar(Foo):
    def properties(self) -> Properties:
        return super().properties()
```

<h3> Repro Steps </h3>

Just put the above code into a single file of its own and open it in VS Code.

<h3> Expected behavior </h3>

The example above has three similar decorators (none of which ever changes the type of the decorated class); it should be possible to use any of them on `Foo` without the linter becoming unhappy.

<h3> Actual behavior </h3>

* With `@decorator1` (active above), I am getting an "class argument must be a base class" linting error.
* With `@decorator2`, the call to `super().properties()` fails; apparently the base class `Foo` is interpreted as just `object`.
* With `@decorator3`, I see the expected behavior.

<h3> Logs </h3>

```
(Client) Pylance async client (2024.5.1) started with python extension (2024.6.0)
[Info  - 13:54:37] (7693) Pylance language server 2024.5.1 (pyright version 1.1.360, commit 4bc8428d) starting
[Info  - 13:54:37] (7693) Server root directory: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist
[Info  - 13:54:37] (7693) Auto-Indent enabled
[Info  - 13:54:37] (7693) Starting service instance "<default>"
(7693) No pyproject.toml file found.
[Info  - 13:54:37] (7693) Setting pythonPath for service "<default>": "/opt/local/bin/python"
[Info  - 13:54:37] (7693) Setting environmentName for service "<default>": "3.11.9 (global)"
[Info  - 13:54:37] (7693) Assuming Python version 3.11.9.final.0
(7693) Assuming Python platform Darwin
[Info  - 13:54:38] (7693) Search paths for file:///%3Cdefault%20workspace%20root%3E
[Info  - 13:54:38] (7693)   /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib
[Info  - 13:54:38] (7693)   /<default workspace root>
[Info  - 13:54:38] (7693)   /<default workspace root>/typings
[Info  - 13:54:38] (7693)   /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stubs/...
[Info  - 13:54:38] (7693)   /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/bundled/stubs
[Info  - 13:54:38] (7693)   /opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11
[Info  - 13:54:38] (7693)   /opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/lib-dynload
[Info  - 13:54:38] (7693)   /Users/hmeine/Library/Python/3.11/lib/python/site-packages
[Info  - 13:54:38] (7693)   /opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages
[Info  - 13:54:38] (7693) Adding fs watcher for library directories:
 file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11
file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/lib-dynload
file:///Users/hmeine/Library/Python/3.11/lib/python/site-packages
file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages
(7693) Searching for source files
(7693) Found module msgpack_numpy : 0.4.8
---- SNIP ----
(7693) Found module zopfli : 0.2.3
[Info  - 13:54:38] (7693) No source files found.
(7693) pytest configurations: {"message":"request cancelled","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
(7693) pytest configurations: {"message":"script","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
(7693) [IDX(FG)] index libraries file:///%3Cdefault%20workspace%20root%3E (index) ...
(7693) [IDX(FG)]   read stdlib indices (7ms)
(7693) [IDX(FG)] index libraries file:///%3Cdefault%20workspace%20root%3E (index) [succeed] (8ms)
[Info  - 13:54:38] (7693) Background analysis(1) root directory: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist
[Info  - 13:54:38] (7693) Background analysis(1) started
(7693) Found module msgpack_numpy : 0.4.8
---- SNIP ----
(7693) Server settings returned for workspace: undefined: {
  "autoSearchPaths": true,
  "disableLanguageServices": false,
  "openFilesOnly": true,
  "useLibraryCodeForTypes": true,
  "watchForSourceChanges": true,
  "watchForLibraryChanges": true,
  "watchForConfigChanges": true,
  "typeCheckingMode": "basic",
  "diagnosticSeverityOverrides": {
    "reportShadowedImports": "warning"
  },
  "logLevel": "log",
  "autoImportCompletions": false,
  "indexing": false,
  "completeFunctionParens": true,
  "enableExtractCodeAction": true,
  "indexOptions": {
    "packageDepths": [
      [
        "sklearn",
        2,
        false
      ],
      [
        "matplotlib",
        2,
        false
      ],
      [
        "scipy",
        2,
        false
      ],
      [
        "django",
        2,
        false
      ],
      [
        "flask",
        2,
        false
      ],
      [
        "fastapi",
        2,
        false
      ]
    ],
    "userFileIndexingLimit": 2000
  },
  "variableInlayTypeHints": false,
  "callArgumentNameInlayHints": "off",
  "functionReturnInlayTypeHints": false,
  "pytestParametersInlayTypeHints": false,
  "importFormat": "absolute",
  "includeFileSpecs": [],
  "excludeFileSpecs": [],
  "ignoreFileSpecs": [],
  "formatOnType": true,
  "taskListTokens": [],
  "enablePytestSupport": true,
  "gotoDefinitionInStringLiteral": true,
  "remapDiagnostics": false,
  "intelliCodeEnabled": true,
  "pythonPath": {
    "_key": "/opt/local/bin/python",
    "_filePath": "/opt/local/bin/python",
    "_query": "",
    "_fragment": "",
    "_originalString": "file:///opt/local/bin/python",
    "_isCaseSensitive": false
  },
  "pythonEnvironmentName": "3.11.9 (global)",
  "fixAll": [],
  "extraCommitChars": false,
  "enableAskingUsersPerformanceReporting": "false",
  "nodeExecutable": "",
  "notebookRunStartupCommands": "",
  "editorIndentConfig": {
    "defaultTabSequence": "    ",
    "tabSize": 4
  }
}
(7693) No pyproject.toml file found.
[Info  - 13:54:38] (7693) Setting pythonPath for service "<default>": "/opt/local/bin/python"
[Info  - 13:54:38] (7693) Setting environmentName for service "<default>": "3.11.9 (global)"
[Info  - 13:54:38] (7693) Assuming Python version 3.11.9.final.0
(7693) Assuming Python platform Darwin
(7693) Found module msgpack_numpy : 0.4.8
(7693) Found module unix_ar : 0.2.1
---- SNIP ----
[Info  - 13:54:38] (7693) Search paths for file:///%3Cdefault%20workspace%20root%3E
[Info  - 13:54:38] (7693)   /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib
[Info  - 13:54:38] (7693)   /<default workspace root>
[Info  - 13:54:38] (7693)   /<default workspace root>/typings
[Info  - 13:54:38] (7693)   /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stubs/...
[Info  - 13:54:38] (7693)   /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/bundled/stubs
[Info  - 13:54:38] (7693)   /opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11
[Info  - 13:54:38] (7693)   /opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/lib-dynload
[Info  - 13:54:38] (7693)   /Users/hmeine/Library/Python/3.11/lib/python/site-packages
[Info  - 13:54:38] (7693)   /opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages
[Info  - 13:54:38] (7693) Adding fs watcher for library directories:
 file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11
file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/lib-dynload
file:///Users/hmeine/Library/Python/3.11/lib/python/site-packages
file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages
(7693) Searching for source files
---- SNIP ----
(7693) pytest configurations: {"message":"request cancelled","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
(7693) pytest configurations: {"message":"spawn /opt/local/bin/python ENOENT","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
(7693) Found module msgpack_numpy : 0.4.8
---- SNIP ----
(7693) pytest configurations: {"message":"spawn /opt/local/bin/python ENOENT","classes":["Test"],"files":["test_*.py","*_test.py"],"functions":["test"]}
(7693) [FG] parsing: file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py (40ms)
(7693) [FG] parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 1ms] (24ms)
(7693) [FG] binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/builtins.pyi (11ms)
(7693) [FG] binding: file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py (1ms)
(7693) [BG(1)] analyzing: file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py ...
(7693) [BG(1)]   parsing: file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py (47ms)
(7693) [BG(1)]   parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 1ms] (24ms)
(7693) [BG(1)]   binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/builtins.pyi (12ms)
(7693) [BG(1)]   binding: file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py (1ms)
(7693) [BG(1)]   checking: file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py ...
(7693) [BG(1)]     parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/typing.pyi [fs read 0ms] (10ms)
(7693) [BG(1)]     binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/typing.pyi (3ms)
(7693) [BG(1)]     parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/types.pyi [fs read 1ms] (4ms)
(7693) [BG(1)]     binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/types.pyi (1ms)
(7693) [BG(1)]     parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi [fs read 1ms] (3ms)
(7693) [BG(1)]     binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi (1ms)
(7693) [BG(1)]     parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/abc.pyi [fs read 1ms] (1ms)
(7693) [BG(1)]     binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/abc.pyi (1ms)
(7693) [BG(1)]     parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/dataclasses.pyi [fs read 1ms] (2ms)
(7693) [BG(1)]     binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/dataclasses.pyi (0ms)
(7693) [BG(1)]     parsing: file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/dataclasses.py [fs read 0ms] (11ms)
(7693) [BG(1)]     binding: file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/dataclasses.py (3ms)
(7693) [BG(1)]     parsing: file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py [fs read 1ms] (21ms)
(7693) [BG(1)]     binding: file:///opt/local/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py (7ms)
(7693) [BG(1)]     parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/typing_extensions.pyi [fs read 0ms] (5ms)
(7693) [BG(1)]     binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/typing_extensions.pyi (1ms)
(7693) [BG(1)]     parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/collections/abc.pyi [fs read 0ms] (0ms)
(7693) [BG(1)]     binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/collections/abc.pyi ...
(7693) [BG(1)]       parsing: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/_collections_abc.pyi [fs read 0ms] (1ms)
(7693) [BG(1)]       binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/_collections_abc.pyi (0ms)
(7693) [BG(1)]     binding: file:///Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2024.5.1/dist/typeshed-fallback/stdlib/collections/abc.pyi (1ms)
(7693) [BG(1)]   checking: file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py (124ms)
[Info  - 13:54:39] (7693) Heap stats: total_heap_size=61MB, used_heap_size=45MB, cross_worker_used_heap_size=45MB, total_physical_size=60MB, total_available_size=4048MB, heap_size_limit=4096MB
(7693) [BG(1)] analyzing: file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py (211ms)
(7693) [BG(1)] SemanticTokens range 0:0 - 32:0 at file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py (1ms)
(7693) [BG(1)] SemanticTokens full at file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py (0ms)
(7693) [BG(1)] SemanticTokens delta previousResultId:1715946879265 at file:///Users/hmeine/Developer/Experiments/pylance_inference_problem.py (0ms)
```

**Comment by erictraut** [comment]
The type annotations in your decorator are incorrect. These decorators are generic and should use a type variable to express this. If you make the following change to your decorators, they will work fine.

```python
def decorator1[T](DataClass: type[T], condition: bool = False) -> type[T]: ...
```

Or if you're using versions of Python prior to 3.12, you can use the old-style TypeVar syntax.

```python
T = TypeVar("T")

def decorator1(DataClass: type[T], condition: bool = False) -> type[T]: ...
```

**Comment by hmeine** [comment]
Thanks for the suggestion. That is good to know, and I already briefly looked into this; unfortunately, I run this code not only with a recent Python but also in an environment that is stuck at Python 3.6, so I hoped that the inferred return type could also be made to work, in particular since it *does* work until I call `dataclasses.is_dataclass()`. Maybe I should try hiding that call from pylance…

**Comment by erictraut** [comment]
@hmeine, Python 3.6 supports type variables. It doesn't support the generic version of `type`, so you'll need to modify my suggested change by swapping out `type[T]` for `Type[T]`.

**Comment by hmeine** [comment]
Thanks a lot.  Indeed that's what I tried after reading your suggestion (and posting my comment – should've edited that), and it works.

Although I had used TypeVar once or twice before and even looked at the typing docs before writing this issue, I could not come up with this myself this time.  I think my problem is that `Type[Base]` means "Base or any type subclassing Base", but the TypeVar suddenly makes the return type say "not *any*, but exactly the *same* type as in the previous instance of the type variable in this signature".  Writing that out, it kind of makes sense, but it was not obvious to me.  While the Python 3.12 syntax is nicer, it still requires the same thinking / interpretation, so I'll have to remember that.

Thanks again – this works perfectly for my environments!
